contributivity_measures.py

```diff
@@ -168,3 +166,704 @@ def not_twice_characteristic(permut):
+        # ## compute some  increments
+        permutation = np.random.permutation(n)
+        for j in range(n):
+            not_twice_characteristic(permutation[: j + 1])
+        permutation = np.flip(permutation)
+        for j in range(n):
+            not_twice_characteristic(permutation[: j + 1])
+        for k in range(n):
+            permutation = np.append(permutation[-1], permutation[:-1])
+            for j in range(n):
+                not_twice_characteristic(permutation[: j + 1])
+
+        # ## do the regressions
+
+        ###### make the datasets
+        def makedata(subset):
+            # compute the size of subset : ||subset||
+            small_node_list = np.array([preprocessed_node_list[i] for i in subset])
+            size_of_S = 0
+            for node in small_node_list:
+                size_of_S += len(node.y_train)
+            data = [size_of_S, size_of_S ** 2] 
+            return data
+
+        datasets = []
+        outputs = []
+        for k in range(n):
+            x = []
+            y = []
+            for subset, incr in increments[k].items():
+                x.append(makedata(subset))
+                y.append(incr)
+            datasets.append(x)
+            outputs.append(y)
+
+        ###### fit the regressions
+        models = []
+        for k in range(n):
+            model_k = LinearRegression()
+            model_k.fit(datasets[k], outputs[k])
+            models.append(model_k)
+
+        # ##define the approximation
+        def approx_increment(subset, k):
+            return models[k].predict([makedata(subset)])[0]
+
+        # compute  the importance density
+        # ## compute the renormalization constant of the importance density for all datatsets
+
+        renorms = []
+        for k in range(n):
+            list_k = np.delete(np.arange(n), k)
+            renorm = 0
+            for length_combination in range(len(list_k) + 1):
+                for subset in combinations(
+                    list_k, length_combination
+                ):  # could be avoided as   prob(np.array(subset))*np.abs(approx_increment(np.array(subset),j)) is constant in the combination
+                    renorm += prob(np.array(subset)) * np.abs(
+                        approx_increment(np.array(subset), k)
+                    )
+            renorms.append(renorm)
+
+        # ## defines the importance density
+        def g(subset, k):
+            return (
+                prob(np.array(subset))
+                * np.abs(approx_increment(np.array(subset), k))
+                / renorms[k]
+            )
+
+        # sampling
+        t = 0
+        q = -norm.ppf((1 - alpha) / 2, loc=0, scale=1)
+        v_max = 0
+        while (
+            t < 100 or t < 4 * q ** 2 * v_max / (sv_accuracy) ** 2
+        ):  # Check if the length of the confidence interval  is below the value of sv_accuracy*characteristic_all_node
+            t += 1
+            print(t)
```

Comment:
The prinds floods the output. We can remove the print or print every 1000th iteration for instance ?